src/edu/washington/escience/myria/perfenforce/PerfEnforceOnlineLearning.java

```diff
@@ -0,0 +1,396 @@
+    int winnerIndex = 0;
+    for (int currentState = 0;
+        currentState < PerfEnforceDriver.configurations.size();
+        currentState++) {
+      double onlinePrediction = queryPredictions[currentState];
+
+      onlinePrediction = (onlinePrediction < 0) ? 0 : onlinePrediction;
+      double currentRatio = 0;
+      if (currentQuery.slaRuntime == 0) {
+        currentRatio = onlinePrediction / 1;
+      } else {
+        currentRatio = onlinePrediction / currentQuery.slaRuntime;
+      }
+      double currentScore = closeToOneScore(currentRatio);
+
+      if (currentScore > maxScore) {
+        winnerIndex = currentState;
+        maxScore = currentScore;
+      }
+    }
+
+    currentConfiguration = PerfEnforceDriver.configurations.get(winnerIndex);
+  }
+
+  /**
+   * Given a configuration size, this method predicts the runtime of the currentQuery
+   *
+   * @param configurationIndex the configuration size in consideration
+   */
+  public void trainOnlineQueries(final int configurationIndex) throws PerfEnforceException {
+    String MOAFileName = Paths.get(onlineLearningPath, "moa.jar").toString();
+    String trainingFileName = Paths.get(onlineLearningPath, "training.arff").toString();
+    String modifiedTrainingFileName =
+        Paths.get(onlineLearningPath, "training-modified-" + configurationIndex + ".arff")
+            .toString();
+    String predictionsFileName =
+        Paths.get(onlineLearningPath, "predictions" + configurationIndex + ".txt").toString();
+
+    try {
+      PrintWriter outputWriter = new PrintWriter(modifiedTrainingFileName);
+      outputWriter.close();
+
+      // copy training file to new file
+      FileChannel src = new FileInputStream(trainingFileName).getChannel();
+      FileChannel dest = new FileOutputStream(modifiedTrainingFileName).getChannel();
+      dest.transferFrom(src, 0, src.size());
+      src.close();
+      dest.close();
+
+      // Append all previous data points
+      FileWriter appendDataWriter = new FileWriter(modifiedTrainingFileName, true);
+      for (String s : previousDataPoints) {
+        appendDataWriter.write(s + "\n");
+      }
+
+      // Append the current point
+      String newPoint = getQueryFeature(configurationIndex, 0);
+      appendDataWriter.write(newPoint + "\n");
+      appendDataWriter.close();
+
+      String moaCommand =
+          String.format(
+              "EvaluatePrequentialRegression -l (rules.functions.Perceptron  -d -l %s) -s (ArffFileStream -f %s) -e (WindowRegressionPerformanceEvaluator -w 1) -f 1 -o %s",
+              .04,
+              modifiedTrainingFileName,
+              predictionsFileName);
+
+      String[] arrayCommand =
+          new String[] {"java", "-classpath", MOAFileName, "moa.DoTask", moaCommand};
+
+      Process p = Runtime.getRuntime().exec(arrayCommand);
+
+      BufferedReader reader = new BufferedReader(new InputStreamReader(p.getInputStream()));
+      while ((reader.readLine()) != null) {}
```

Comment:
Yes, I ran into an issue where the process would just hang even when I tried to flush the error stream. I ended up following suggestions from http://stackoverflow.com/questions/5483830/process-waitfor-never-returns.